src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -208,39 +208,64 @@ public void OnRenderCompleted(long incomingBatchId, string errorMessageOrNull)
                 return;
             }
 
+            // When clients send acks we know for sure they received and applied the batch.
+            // We send batches right away, and hold them in memory until we receive an ACK.
+            // If one or more client ACKs get lost (e.g., with long polling, client->server delivery is not guaranteed) we might receive an ack for a higher batch.
+            // We confirm all previous batches at that point (because receiving an ack is guarantee
+            // from the client that it has received and successfully applied all batches up to that point).
+
+            // If receive an ack for a previously acknowledged batch, its an error, as the messages are
+            // guranteed to be delivered in order, so a message for a render batch of 2 will never arrive
+            // after a message for a render batch for 3.
+            // If that were to be the case, it would just be enough to relax the checks here and simply skip
+            // the message.
+
+            // A batch might get lost when we send it to the client, because the client might disconnect before receiving and processing it.
+            // In this case, once it reconnects the server will re-send any unacknowledged batches, some of which the
+            // client might have received and even believe it did send back an acknowledgement for. The client handles
+            // those by re-acknowledging.
+            // it was expecting a previous batch, it can simply force a reconnection.
+
             // Even though we're not on the renderer sync context here, it's safe to assume ordered execution of the following
             // line (i.e., matching the order in which we received batch completion messages) based on the fact that SignalR
             // synchronizes calls to hub methods. That is, it won't issue more than one call to this method from the same hub
             // at the same time on different threads.
-            if (!PendingRenderBatches.TryDequeue(out var entry))
+
+            if (!PendingRenderBatches.TryPeek(out var nextUnacknowledgedBatch) || incomingBatchId < nextUnacknowledgedBatch.BatchId)
             {
-                HandleException(
-                    new InvalidOperationException($"Received a notification for a rendered batch when not expecting it. Batch id '{incomingBatchId}'."));
+                Log.ReceivedDuplicateBatchAck(_logger, incomingBatchId);
             }
             else
             {
-                entry.Data.Dispose();
```

Comment:
Where did this go?

Reply:
I thought we needed this because we were dequeuing the entry before we could know we could get rid of it. Is that not the case?

Shouldn't we have a test for this? I ran all the tests and nothing failed 😕

Reply:
I re-added the check